Added support for quantization parameters in RKNN model conversion.

BlackHeartZero opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Add configurable RKNN quantization support to the exporter, enabling dataset-driven quantization and algorithm/dtype selection during RKNN model conversion. 🤖⚙️

### 📊 Key Changes
- Introduces new RKNN export args in default.yaml: `qt_datasets`, `do_quantization`, `quantized_dtype`, `quantized_algorithm`.
- Updates export_formats() to surface the new RKNN arguments in the CLI/help matrix.
- Enhances `export_rknn` to:
  - Pass `quantized_dtype` and `quantized_algorithm` to `rknn.config`.
  - Enable optional quantization via `rknn.build(do_quantization=..., dataset=...)` using the provided dataset file.
  - Create a dedicated output directory and write metadata as before.

### 🎯 Purpose & Impact
- Adds first-class quantization control for RKNN exports, improving deployment on Rockchip devices. 📱🧠
- Allows users to choose quantization dtype/algorithm and supply calibration datasets for better accuracy/size trade-offs.
- Potential behavior change: `do_quantization` defaults to True—without `qt_datasets` provided, builds may fail or differ; consider verifying dataset availability in usage/docs. 🔎

Y-T-G committed:
Fix undefined `final_epoch` for DDP training when `val=False` (#22607)

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>
Co-authored-by: UltralyticsAssistant <web@ultralytics.com>

BlackHeartZero committed:
Added support for quantization parameters in RKNN model conversion.

UltralyticsAssistant commented:
👋 Hello @BlackHeartZero, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review assistant; an Ultralytics engineer will follow up to help finalize your contribution.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any relevant issues (https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics Continuous Integration (CI) (https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant documentation (https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our Contributor License Agreement (https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our Contributing Guide (https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please gate RKNN quantization behind the presence of a dataset so the default configuration keeps working without raising RKNN build errors.

💬 Posted 1 inline comment

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

BlackHeartZero commented:
I have read the CLA Document and I sign the CLA

BlackHeartZero commented on ultralytics/engine/exporter.py:
```diff
@@ -1157,11 +1157,17 @@ def export_rknn(self, prefix=colorstr("RKNN:")):
         rknn.load_onnx(model=f)
-        rknn.build(do_quantization=False)  # TODO: Add quantization support
+        # Use qt_datasets if provided, otherwise use default dataset file
+        
```
i have got the suggest

BlackHeartZero commented:
I have read the CLA Document and I sign the CLA

glenn-jocher committed:
Python and URL/301 refactor (#22618)

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>
Co-authored-by: UltralyticsAssistant <web@ultralytics.com>

glenn-jocher committed:
Fix `check_requirements()` missing output when using `uv` package manager (#22623)

onuralpszr committed:
`ultralytics 8.3.226` ♻️ Replace `eval()` with `ast.literal_eval()` for security (#22597)

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>
Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>
Co-authored-by: UltralyticsAssistant <web@ultralytics.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Add ConsoleLogger docstrings (#22624)

Co-authored-by: UltralyticsAssistant <web@ultralytics.com>

glenn-jocher committed:
Change Dependabot update schedule to daily (#22631)

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

BlackHeartZero commented:
@glenn-jocher  I believe quantized_dtype and quantized_algorithm are important and necessary parameters. I hope we can finalize where to place them soon.

glenn-jocher commented:
@BlackHeartZero we're working on a better generalized argument passing approach for methods (i.e. `model.export()`), but for now we're seeking to not expand default.yaml as it's getting too complicated.

For the duplicated args please use `data`, `float` and `int8` and for the other specific ones these are the export args currently available:
https://docs.ultralytics.com/modes/export/#arguments

onuralpszr committed:
fix: 🐞 Update `scipy-stubs` dependency to require version `>=1.14.1.4` for Python 3.10+ (#22634)

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

RizwanMunawar committed:
Update `citation` of `african-wildlife` dataset (#22632)

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>
Co-authored-by: UltralyticsAssistant <web@ultralytics.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

onuralpszr committed:
chore: 🧹 Add video and image file extensions to .gitignore (#22635)

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update Edge TPU installation for keyring (#22638)

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

RizwanMunawar committed:
`ultralytics 8.3.227` Pin `onnx<=1.19.1` for Conda compatibility (#22628)

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
New Ultralytics `chat.js` bot (#22639)

BlackHeartZero committed:
Refactor RKNN export to use existing parameters

- Use existing 'data' parameter instead of 'qt_datasets'
- Use existing 'int8' parameter instead of 'do_quantization'
- Hard-code additional RKNN parameters with common defaults
- Update RKNN documentation with new parameter usage
- Add test script for RKNN export functionality

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr commented:
Closing in favor new PR cases discussion: https://github.com/ultralytics/ultralytics/pull/24384 (this one very old PR so resolve conflict more time vs new PR) new PR would be nice. 

cc @lakshanthad

glenn-jocher commented:
Closing the loop here: RKNN INT8 export is now supported in Ultralytics starting with `ultralytics>=8.4.59`.

Implementation PR: https://github.com/ultralytics/ultralytics/pull/24656
Platform export UI: https://platform.ultralytics.com/ultralytics/yolo26/yolo26n?tab=export

Python usage now follows the same INT8 calibration-data flow as other export formats, for example:

```python
from ultralytics import YOLO

YOLO("yolo26n.pt").export(format="rknn", name="rk3588", int8=True, data="coco8.yaml")
```

Platform also exposes the RKNN FP16/INT8 export controls directly:

<img width="1306" height="1036" alt="Screenshot 2026-06-01 at 17 31 16" src="https://github.com/user-attachments/assets/553404ae-3048-48f9-9fde-02b200991be2" />
